#! /bin/sh /usr/share/dpatch/dpatch-run
##
## DP: Addes an error indicator to the recording info, which counts
## DP: the number of discontinuity errors

@DPATCH@
diff --git a/recorder.c b/recorder.c
index 9147e2a..877be30 100644
--- a/recorder.c
+++ b/recorder.c
@@ -121,6 +121,7 @@ void cRecorder::Action(void)
   time_t t = time(NULL);
   bool InfoWritten = false;
   bool FirstIframeSeen = false;
+  cTsStreamErrorCounter TsStreamErrorCounter;
   while (Running()) {
         int r;
         uchar *b = ringBuffer->Get(r);
@@ -156,6 +157,16 @@ void cRecorder::Action(void)
                              fileSize += TS_SIZE;
                              }
                        }
+                    TsStreamErrorCounter.CheckTsPackets(b, Count);
+                    if (InfoWritten && TsStreamErrorCounter.ErrorCount() > 0) {
+                       cRecordingInfo RecordingInfo(recordingName);
+                       if (RecordingInfo.Read()) {
+                          RecordingInfo.SetErrorCount(RecordingInfo.ErrorCount() + TsStreamErrorCounter.ErrorCount());
+                          RecordingInfo.Write();
+                          Recordings.UpdateByName(recordingName);
+                          TsStreamErrorCounter.ResetErrorCount();
+                          }
+                       }
                     if (recordFile->Write(b, Count) < 0) {
                        LOG_ERROR_STR(fileName->Name());
                        break;
diff --git a/recording.c b/recording.c
index 327a300..7445279 100644
--- a/recording.c
+++ b/recording.c
@@ -317,6 +317,7 @@ cRecordingInfo::cRecordingInfo(const cChannel *Channel, const cEvent *Event)
   event = ownEvent ? ownEvent : Event;
   aux = NULL;
   framesPerSecond = DEFAULTFRAMESPERSECOND;
+  errorCount = 0;
   priority = MAXPRIORITY;
   lifetime = MAXLIFETIME;
   fileName = NULL;
@@ -410,6 +411,11 @@ void cRecordingInfo::SetFramesPerSecond(double FramesPerSecond)
   framesPerSecond = FramesPerSecond;
 }
 
+void cRecordingInfo::SetErrorCount(int ErrorCount)
+{
+  errorCount = ErrorCount;
+}
+
 bool cRecordingInfo::Read(FILE *f)
 {
   if (ownEvent) {
@@ -449,6 +455,8 @@ bool cRecordingInfo::Read(FILE *f)
                        break;
              case 'F': framesPerSecond = atof(t);
                        break;
+             case 'R': errorCount = atoi(t);
+                       break;
              case 'L': lifetime = atoi(t);
                        break;
              case 'P': priority = atoi(t);
@@ -475,6 +483,7 @@ bool cRecordingInfo::Write(FILE *f, const char *Prefix) const
      fprintf(f, "%sC %s%s%s\n", Prefix, *channelID.ToString(), channelName ? " " : "", channelName ? channelName : "");
   event->Dump(f, Prefix, true);
   fprintf(f, "%sF %.10g\n", Prefix, framesPerSecond);
+  fprintf(f, "%sR %d\n", Prefix, errorCount);
   fprintf(f, "%sP %d\n", Prefix, priority);
   fprintf(f, "%sL %d\n", Prefix, lifetime);
   if (aux)
diff --git a/recording.h b/recording.h
index 34cccac..97c9425 100644
--- a/recording.h
+++ b/recording.h
@@ -54,6 +54,7 @@ private:
   cEvent *ownEvent;
   char *aux;
   double framesPerSecond;
+  int errorCount;
   int priority;
   int lifetime;
   char *fileName;
@@ -73,6 +74,8 @@ public:
   const char *Aux(void) const { return aux; }
   double FramesPerSecond(void) const { return framesPerSecond; }
   void SetFramesPerSecond(double FramesPerSecond);
+  int ErrorCount(void) const { return errorCount; }
+  void SetErrorCount(int ErrorCount);
   bool Read(FILE *f);
   bool Write(FILE *f, const char *Prefix = "") const;
   bool Read(void);
diff --git a/remux.c b/remux.c
index f7ad86d..abae484 100644
--- a/remux.c
+++ b/remux.c
@@ -970,3 +970,44 @@ int cFrameDetector::Analyze(const uchar *Data, int Length)
         }
   return Processed;
 }
+
+// --- cTsStreamErrorCounter -------------------------------------------------
+
+cTsStreamErrorCounter::cTsStreamErrorCounter(void)
+{
+  errorCount = 0;
+  for(int i=0; i < MAX_PIDS; i++) pids[i] = -1;
+  memset(counters, 0, sizeof(counters));
+}
+
+void cTsStreamErrorCounter::CheckTsPackets(uchar* Data, int Length)
+{
+  while (Length >= TS_SIZE) {
+        int pid = ((Data[1] & 0x1F) << 8) + Data[2];
+        char ctr = (Data[3] & 0x0F);
+        char adaption = (Data[3] & 0x30) >> 4;
+
+        if (adaption != 2)
+           CheckTsPacketContinuity(pid, ctr);
+
+        Length -= TS_SIZE;
+        Data += TS_SIZE;
+       }
+}
+
+void cTsStreamErrorCounter::CheckTsPacketContinuity(int pid, int counter)
+{
+  for (int i=0; i < MAX_PIDS; i++) {
+      if (pids[i] == -1) {
+         pids[i] = pid;
+         counters[i] = counter;
+         break;
+         }
+      if (pids[i] == pid) {
+         if (((counter - counters[i]) & 0x0F) != 1)
+            errorCount++;
+         counters[i] = counter;
+         break;
+         }
+      }
+}
diff --git a/remux.h b/remux.h
index 7140aaf..78025dc 100644
--- a/remux.h
+++ b/remux.h
@@ -383,4 +383,29 @@ public:
       ///< available.
   };
 
+
+// TS stream error counter:
+// Count (continuity) errors in a TS packet stream
+
+#define MAX_PIDS 20
+
+class cTsStreamErrorCounter {
+private:
+  int errorCount;
+  int pids[MAX_PIDS];
+  uchar counters[MAX_PIDS];
+public:
+  cTsStreamErrorCounter(void);
+  void CheckTsPackets(uchar* Data, int Length);
+      ///< Check for errors in the TS packets pointed to by Data. Length is the
+      ///< number of bytes Data points to, and must be a multiple of 188.
+  int ErrorCount(void) const { return errorCount; };
+      ///< Return the number of errors found in the analyzed TS packet stream.
+      ///< Right now, only continuity errors are reported.
+  void ResetErrorCount(void) { errorCount = 0; };
+      ///< Reset error count to 0
+private:
+  void CheckTsPacketContinuity(int pid, int counter);
+  };
+
 #endif // __REMUX_H
-- 
